imports/wasi_snapshot_preview1/fs.go

```diff
@@ -643,11 +643,245 @@ func fdRead_shouldContinueRead(n, l uint32, err error) (bool, Errno) {
-)
+var fdReaddir = &wasm.HostFunc{
+	ExportNames: []string{functionFdReaddir},
+	Name:        functionFdReaddir,
+	ParamTypes:  []wasm.ValueType{i32, i32, i32, i64, i32},
+	ParamNames:  []string{"fd", "buf", "buf_len", "cookie", "result.bufused"},
+	ResultTypes: []api.ValueType{i32},
+	Code: &wasm.Code{
+		IsHostFunction: true,
+		GoFunc:         wasiFunc(fdReaddirFn),
+	},
+}
+
+func fdReaddirFn(ctx context.Context, mod api.Module, params []uint64) Errno {
+	fd := uint32(params[0])
+	buf := uint32(params[1])
+	bufLen := uint32(params[2])
+	cookie := params[3]
+	resultBufused := uint32(params[4])
+
+	// Validate the FD is a directory
+	rd, dir, errno := openedDir(ctx, mod, fd)
+	if errno != ErrnoSuccess {
+		return errno
+	}
+
+	// expect a cookie only if we are continuing a read.
+	if cookie == uint64(0) && dir.CountRead > 0 {
+		return ErrnoInval // invalid as a cookie is minimally one.
+	}
+
+	// Ensure we have the max directory entries that can be serialized as
+	// dirents, knowing file names come after.
+	maxDirEntries := int(bufLen/direntSize + 1)
+
+	// Add one more to know if maxDirEntries is the end of the directory.
+	//	>> If less than the size of the read buffer, the end of the
+	//	>> directory has been reached.
+	maxDirEntries += 1
+
+	// Entries here are truncated to those that start at the cookie.
+	entries, errno := lastDirEntries(dir, cookie)
+	if errno != ErrnoSuccess {
+		return errno
+	}
+
+	// Append more entries as needed to fill bufLen
+	if entryCount := len(entries); entryCount < maxDirEntries {
+		if l, err := rd.ReadDir(maxDirEntries - entryCount); err != io.EOF {
+			if err != nil {
+				return ErrnoIo
+			}
+			dir.CountRead += uint64(len(l))
+			entries = append(entries, l...)
+			// Replace the cache with up to maxDirEntries, starting at cookie.
+			dir.Entries = entries
+		}
+	}
+
+	mem := mod.Memory()
+
+	// Determine how many dirents we can write, excluding a potentially
+	// truncated entry.
+	bufused, direntCount, truncatedEntryLen := maxDirents(entries, bufLen)
+
+	// Now, write entries to the underlying buffer.
+	if bufused > 0 {
+
+		// Cookie is zero when there is no cookie. Now that we know we will
+		// write entries, ensure the first one is one not zero.
+		if cookie == uint64(0) {
+			cookie = 1
+		}
+
+		dirents, ok := mem.Read(ctx, buf, bufused)
+		if !ok {
+			return ErrnoFault
+		}
+
+		writeDirents(entries, direntCount, truncatedEntryLen, dirents, cookie)
+	}
+
+	if !mem.WriteUint32Le(ctx, resultBufused, bufused) {
+		return ErrnoFault
+	}
+	return ErrnoSuccess
+}
+
+// lastDirEntries is broken out from fdReaddirFn for testability.
+func lastDirEntries(dir *internalsys.ReadDir, cookie uint64) (entries []fs.DirEntry, errno Errno) {
+	entryCount := len(dir.Entries)
+	if entryCount == 0 { // there was no prior call
+		if cookie != 0 {
+			errno = ErrnoInval // invalid as we haven't sent that cookie
+		}
+		return
+	}
+
+	countRead := int(dir.CountRead)
+	if int(cookie)-countRead == 1 {
```

Comment:
cookie is 64-bit so there's a possibility `int(cookie)` can be negative, though I'm not sure if we want to care about that :D

Reply:
maybe we can change the arg type to int64

Reply:
made cookie signed